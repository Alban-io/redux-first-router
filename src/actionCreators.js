export const INIT = '@@address-bar/INIT' //used internally by middleware and then action is re-written
export const NOT_FOUND = '@@address-bar/NOT_FOUND'


/** EXPORTED INIT ACTION CREATORS: 
 *  for use when ready key is not provided and 
 *  manual control of initialization is desired
*/

export function init(pathname) {
  return {
    type: INIT,
    payload: {
      pathname,
    }
  }
}

export function initThunk() {
  return (dispatch, getState) => {
    let {pathname} = getState().location
    return dispatch(init(pathname))
  }
}


/** PRIMARY/SOLE NAVIGATION ACTION CREATOR 
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'