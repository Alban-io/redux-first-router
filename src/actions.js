export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

/** PRIMARY/SOLE NAVIGATION ACTION CREATOR
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'
