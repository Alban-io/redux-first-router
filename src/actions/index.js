export { default as redirect } from './redirect'
export { default as notFound } from './notFound'
export { default as addRoutes } from './addRoutes'
export { default as changeBasename } from './changeBasename'

export {
  push,
  replace,
  setState,
  jump,
  back,
  next,
  reset
} from './history'
