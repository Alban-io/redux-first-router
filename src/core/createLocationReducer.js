// @flow
import { ADD_ROUTES } from '../types'
import { isServer, urlToAction, typeToScene } from '../utils'

import type {
  LocationState,
  RoutesMap,
  Action,
  History
} from '../flow-types'

export default (routes: RoutesMap, history: History) => {
  const initialState = createInitialState(routes, history)

  return (
    st: LocationState = initialState,
    action: Action
  ): LocationState => {
    const r = routes[action.type]
    const l = action.location

    if (r && r.path && (
      l.url !== st.url
      || (l.basename && l.basename !== st.basename)
      || l.kind === 'load'
      || action.info === 'reset')
    ) {
      const { type, params, query, state, hash } = action
      return { type, params, query, state, hash, hasSSR: st.hasSSR, ...l }
    }
    else if (action.type === ADD_ROUTES) {
      const count = Object.keys(action.payload.routes).length  // we need to be able to update Links when new routes are added
      const routesAdded = (st.routesAdded || 0) + count        // we could just increment a number, but why not at least off some info
      return { ...st, routesAdded }
    }
    else if (l && l.kind === 'setState') {
      const { state, location: { entries } } = action
      return { ...st, state, entries }
    }
    else if (action.type.indexOf('_ERROR') > -1) {
      const { error, type: errorType } = action
      return { ...st, error, errorType }
    }

    return st
  }
}

export const createInitialState = (
  routes: RoutesMap,
  history: History
): LocationState => {
  const { kind, entries, index, length, location } = history
  const { url, pathname, search, basename } = location
  const action = urlToAction(location, routes)
  const { type, params = {}, query = {}, state = {}, hash = '' } = action
  const scene = typeToScene(type)
  const hasSSR = isServer()

  return {
    type,
    params,
    query,
    hash,
    state,

    url,
    pathname,
    search,
    basename,
    scene,

    kind: 'init',
    entries,
    index,
    length,

    hasSSR,

    prev: createPrev(hasSSR)
  }
}

export const createPrev = (hasSSR: boolean) => ({
  type: '',
  params: {},
  query: {},
  hash: '',
  state: {},

  url: '',
  pathname: '',
  search: '',
  basename: '',
  scene: '',

  kind: '',
  entries: [],
  index: -1,
  length: 0,

  hasSSR
})
