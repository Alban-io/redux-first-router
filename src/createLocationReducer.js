// @flow
import { NOT_FOUND, ADD_ROUTES } from './index'
import isServer from './utils/isServer'
import pathToAction from './utils/pathToAction'

import type {
  LocationState,
  RoutesMap,
  Action,
  Options,
  History,
  ReceivedAction
} from './flow-types'

export default (routes: RoutesMap, history: History, options: Options) => {
  const initialState = createInitialState(routes, history, options)

  return (
    state: LocationState = initialState,
    action: Action
  ): LocationState => {
    const route = routes[action.type]

    if (
      action.type === NOT_FOUND ||
      (route &&
        !action.error &&
        (typeof route === 'string' || route.path) &&
        (action.meta.location.current.pathname !== state.pathname ||
          action.meta.location.current.search !== state.search ||
          action.meta.location.kind === 'load'))
    ) {
      const query = action.meta.location.current.payload.query
      const search = action.meta.location.current.search
      const payload = action.payload || {}
      delete payload.query

      return {
        pathname: action.meta.location.current.pathname,
        type: action.type,
        payload,
        ...(query && { query, search }),
        prev: action.meta.location.prev,
        kind: action.meta.location.kind,
        basename: action.meta.location.basename,
        entries: action.meta.location.history.entries,
        index: action.meta.location.history.index,
        length: action.meta.location.history.length,
        hasSSR: state.hasSSR
      }
    }
    else if (action.type === ADD_ROUTES) {
      const count = Object.keys(action.payload.routes).length     // we need to be able to update Links when new routes are added
      const routesAdded = (state.routesAdded || 0) + count        // we could just increment a number, but this is more informative
      return { ...state, routesAdded }
    }
    else if (action.type.indexOf('ERROR') > -1) {
      const { error, type: errorType } = action
      return { ...state, error, errorType }
    }

    return state
  }
}

export const createInitialState = (
  routes: RoutesMap,
  history: History,
  options: Options
): LocationState => {
  const path = history.location.url
  const initialAction = pathToAction(path, routes, options.basename)
  const { type, payload, meta }: ReceivedAction = initialAction

  return {
    pathname: path.split('?')[0],
    type,
    payload,
    ...meta,
    prev: {
      pathname: '',
      type: '',
      payload: {},
      kind: '',
      index: -1,
      length: 0
    },
    kind: 'init',
    entries: history.entries,
    index: history.index,
    length: history.length,
    hasSSR: isServer() ? true : undefined // client uses initial server `hasSSR` state setup here
  }
}
