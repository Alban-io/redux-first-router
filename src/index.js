// @flow
export const PREFIX = '@@rudy'
export const prefixType = (type: string) => `${PREFIX}/${type}`

export const UPDATE_HISTORY = prefixType('UPDATE_HISTORY')
export const NOT_FOUND = prefixType('NOT_FOUND')
export const ADD_ROUTES = prefixType('ADD_ROUTES')
export const COMPLETE = prefixType('COMPLETE')
export const ERROR = prefixType('ERROR')

export { default as createRouter } from './createRouter'

export { default as redirect } from './action-creators/redirect'
export { default as notFound } from './action-creators/notFound'
export { default as addRoutes } from './action-creators/addRoutes'
export { default as changeBasename } from './action-creators/changeBasename'

import * as history from './action-creators/history'
export { history }

export { default as actionToPath } from './utils/actionToPath'
export { default as pathToAction } from './utils/pathToAction'
export { default as isLocationAction } from './utils/isLocationAction'
export { default as doesRedirect } from './utils/doesRedirect'


export type {
  RouteInput,
  Route,
  RoutesMap,
  Routes,
  RouteNames,
  Options,
  Params,
  Payload,
  LocationState,
  Location,
  Meta,
  Action,
  ReceivedAction,
  Listener,
  Listen,
  Push,
  GoBack,
  History,
  HistoryLocation,
  Document,
  Navigators,
  Navigator,
  Store,
  NavigationAction,
  Router
} from './flow-types'
