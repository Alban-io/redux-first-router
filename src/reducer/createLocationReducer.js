// @flow
import { NOT_FOUND } from '../index'
import isServer from '../pure-utils/isServer'
import type { LocationState, RoutesMap, Action, Payload, History } from '../flow-types'


export default (
  initialState: LocationState,
  routesMap: RoutesMap,
) => (
  state: LocationState = initialState,
  action: Action,
): LocationState => {
  if (action.type === NOT_FOUND || (routesMap[action.type]
      && (action.meta.location.current.pathname !== state.pathname || action.meta.location.load))) {
    return {
      pathname: action.meta.location.current.pathname,
      type: action.type,
      payload: { ...action.payload },
      prev: action.meta.location.prev,
      load: action.meta.location.load,
      backNext: action.meta.location.backNext,
      redirect: action.meta.location.redirect,
      history: action.meta.location.history,
      hasSSR: state.hasSSR
      routesMap
    }
  }

  return state
}


export const getInitialState = (
  currentPathname: string,
  type: string,
  payload: Payload,
  routesMap: RoutesMap,
  history: History,
): LocationState => ({
  pathname: currentPathname,
  type,
  payload,
  prev: {
    pathname: '',
    type: '',
    payload: {},
  },
  load: undefined,
  backNext: undefined,
  redirect: undefined,
  history: !history.entries ? undefined : {
    entries: history.entries.map(entry => entry.pathname),
    index: history.index,
    length: history.length,
  },
  hasSSR: isServer() ? true : undefined,
  routesMap,
})

