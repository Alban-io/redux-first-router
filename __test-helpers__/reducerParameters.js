import createSmartHistory from '../src/smart-history'

import { createInitialState } from '../src/createLocationReducer'

export default async (type, pathname) => {
  // eslint-disable-line import/prefer-default-export
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  await history.push(pathname)

  const current = { pathname, type, payload: { param: 'bar' } }
  const prev = { pathname: '/first', type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: { path: '/first' },
    SECOND: { path: '/second/:param' }
  }

  return {
    type,
    pathname,
    current,
    prev,

    initialState: createInitialState(
      routesMap,
      history
    ),

    routesMap,
    history,

    action: {
      type,
      payload: { param: 'bar' },
      meta: {
        location: {
          current,
          prev,
          kind: 'load',
          history: {
            entries: history.entries.slice(0), // history.entries.map(entry => entry.pathname)
            index: history.index,
            length: history.length
          }
        }
      }
    },

    expectState(state) {
      expect(state.pathname).toEqual(pathname)
      expect(state.type).toEqual(type)
      expect(state.payload).toEqual({ param: 'bar' })
      expect(state.prev).toEqual(prev)
      expect(state.kind).toEqual('load')

      expect(state).toMatchSnapshot()
    }
  }
}
