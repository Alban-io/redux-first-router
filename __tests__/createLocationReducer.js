import createSmartHistory from '../src/smart-history'
import createLocationReducer, {
  createInitialState
} from '../src/createLocationReducer'
import { NOT_FOUND } from '../src/index'
import reducerParameters from '../__test-helpers__/reducerParameters'

it('createLocationReducer() - maintains address bar pathname state and current + previous matched location-aware actions', async () => {
  const { history, routesMap, action, expectState } = await reducerParameters(
    'SECOND',
    '/second/bar'
  )

  const reducer = createLocationReducer(routesMap, history, {})
  const state = reducer(undefined, action) /*? */

  expectState(state)
})

it('locationReducer() reduces action.type === NOT_FOUND', async () => {
  const { history, routesMap, action } = await reducerParameters(
    NOT_FOUND,
    '/foo'
  )

  const reducer = createLocationReducer(routesMap, history, {})
  const state = reducer(undefined, action) /*? */

  expect(state.type).toEqual(NOT_FOUND)
  expect(state.pathname).toEqual('/foo')
})

it('locationReducer() reduces non matched action.type and returns initialState', async () => {
  const { initialState, history, routesMap, action } = await reducerParameters(
    'THIRD',
    '/third'
  )

  const reducer = createLocationReducer(routesMap, history, {})
  const state = reducer(undefined, action) /*? */

  expect(state).toEqual(initialState)
})

it('getInitialState() returns state.history === undefined when using createBrowserHistory', async () => {
  const pathname = '/first'
  const history = createSmartHistory({ initialEntries: ['/first'] })
  history.listen(function() {})
  const current = { pathname, type: 'FIRST', payload: {} }
  const routesMap = {
    FIRST: '/first'
  }

  history.entries = undefined
  const initialState = createInitialState(routesMap, history, {})

  expect(initialState.history).not.toBeDefined()
  expect(initialState).toMatchSnapshot()
})
