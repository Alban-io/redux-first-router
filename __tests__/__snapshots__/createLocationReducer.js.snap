// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLocationReducer() - maintains address bar pathname state and current + previous matched location-aware actions 1`] = `
Object {
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
  ],
  "hasSSR": false,
  "hash": undefined,
  "index": 0,
  "kind": "load",
  "length": 1,
  "pathname": "/second/bar",
  "payload": Object {
    "param": "bar",
  },
  "prev": Object {
    "pathname": "/first",
    "payload": Object {},
    "type": "FIRST",
  },
  "query": undefined,
  "state": undefined,
  "type": "SECOND",
  "url": "/second/bar",
}
`;

exports[`getInitialState() returns state.history === undefined when using createBrowserHistory 1`] = `
Object {
  "basename": "",
  "entries": undefined,
  "hasSSR": false,
  "hash": "",
  "index": 0,
  "kind": "init",
  "length": 1,
  "pathname": "/first",
  "payload": Object {},
  "prev": Object {
    "basename": "",
    "entries": Array [],
    "hasSSR": false,
    "hash": "",
    "index": -1,
    "kind": "",
    "length": 0,
    "pathname": "",
    "payload": Object {},
    "query": Object {},
    "scene": "",
    "search": "",
    "state": Object {},
    "type": "",
    "url": "",
  },
  "query": Object {},
  "scene": "",
  "search": "",
  "state": Object {},
  "type": "FIRST",
  "url": "/first",
}
`;
